### PR TITLE
Convert plugins.jenkins.io to a static compiled html site

### PR DESCRIPTION
**Pull request by halkeye**
## Changes
* Switched from really old nextjs to gatsby
* Upgraded React
* Removed redux (gatsby data layer takes care of the majority, and searching is just an api call)
* Cleaned up some of the CSS and switched to more bootstrap friendly containers, so its more mobile friendly

## Outstanding issues

- [x] Remove Fake dialog box
- [x] upgrade everything to node 13
- [X] Figure out deployment (For now I'm going to split the codebase into 2 repos.  1) plugin-site-api docker image (blocked by https://issues.jenkins-ci.org/browse/INFRA-2400)  2) Update helm chart to deploy both docker images independantly using the same ingress)
- [X] Revert jenkinsfile to something that runs in ci.jenkins
- [x] > The icon view / detail view / list view switch on the search page doesn't work and is rendered strangely
- [x] change urls to plugins.jenkins.io / make overrideable by env variables
- [x] The category links don’t work on the home page
- [x] Browse/search results page is missing title
- [x] Still says `Jenkins Plugins | Jenkins Plugins` in various places
- [ ] Adding new 

## Delaying till later

- [ ] Going to page 2 of search results, then searching for more specific (adding labels/categorys/or query string) will show no results cause on page 2 not page 1
- [ ] More mobile fixes
- [ ] Completely remove the main.module.css?
- [ ] Remove runtime dependancy on api?

## Demo / See it in glorious action:
https://jenkins-plugins.g4v.dev/

### CODE REVIEW

**Comment by rtyler** [comment]
@halkeye Interesting work, the site looks promising but what does this mean "staticly compile?" Does this remove the need for the API server behind the scenes?

**Comment by halkeye** [comment]
It uses the API server for search, and the initial compiling.

The search could easily be switched away from API server.

I think lots of the API server functionality could get moved to update center.

**Comment by rtyler** [comment]
@halkeye Yeah, I have long thought about just removing the API server from the equation and just grabbing the update center for the page to use at runtime.

The Update Center JSON is a little fat, but it would allow plugins.jenkins.io to be run off github pages entirely

**Comment by halkeye** [comment]
> The Update Center JSON is a little fat, but it would allow plugins.jenkins.io to be run off github pages entirely

It doesn't have to have the content inside the update center json, it could easily be seperate files

Then different things could start consuming more and more of it, like inside jenkins could show the readme as an iframe or something.

> Yeah, I have long thought about just removing the API server from the equation and just grabbing the update center for the page to use at runtime.

I'm pretty sure it would only take me a couple hours at most to remove API from runtime for search, i did it because I thought "relevance" search was something special.

**Comment by halkeye** [comment]
rebased on origin/master, so when the time comes its mergable

**Comment by timja** [comment]
just had a look at mobile, I think if we remove the padded class next to the `container-fluid` or just it so it only applies on desktop it looks a lot better

image

File: gatsby-config.js
Comment (timja) [inline gatsby-config.js]: needs to be configurable or hardcoded to plugins.jenkins.io

File: plugins/gatsby-source-jenkinsplugins/.gitignore
Comment (timja) [inline plugins/gatsby-source-jenkinsplugins/.gitignore]: this file needed?

File: plugins/gatsby-source-jenkinsplugins/.npmignore
Comment (timja) [inline plugins/gatsby-source-jenkinsplugins/.npmignore]: is this needed as well?

File: plugins/gatsby-source-jenkinsplugins/gatsby-node.js
Comment (timja) [inline plugins/gatsby-source-jenkinsplugins/gatsby-node.js]: should this be a do while loop instead? there's a condition that breaks out of this loop

File: src/components/Icon.jsx
Comment (timja) [inline src/components/Icon.jsx]: is this duplicated from that helper? not ideal

File: src/components/Pages.jsx
Comment (timja) [inline src/components/Pages.jsx]: ?

File: src/components/Pages.jsx
Comment (timja) [inline src/components/Pages.jsx]: ?

File: src/styles/base.css
Comment (timja) [inline src/styles/base.css:870]: do we need it always?

File: utils.js
Comment (timja) [inline utils.js]: couldn't we just add all this with a multi line string?

File: README.md
Comment (timja) [inline README.md]: node 12 / 13?

File: README.md
Comment (timja) [inline README.md:10]: is webpack still in use?

**Review by timja — COMMENTED** [review]
Few comments but looking great,
I may have missed it but what's the plan for updating the content?

**Comment by timja** [comment]
There's a deprecation warning when running `yarn dev`:
```
boundActionCreators is deprecated. Please use actions instead. For migration instructions, see https://gatsby.dev/boundActionCreators
Check the following files:
⠀
dist/client/main.js
```

**Comment by timja** [comment]
couple of errors in the console
image

**Comment by timja** [comment]
How do I get the content to show up locally?

I tried with `SLOW_MODE`  the only way I could get that to even finish was to add a hosts file entry, otherwise it kept crashing with random unknown hosts after a few minutes.

But even then there was no content

File: plugins/gatsby-source-jenkinsplugins/.npmignore
Comment (halkeye) [reply plugins/gatsby-source-jenkinsplugins/.npmignore]: Nope, left behind files from the source I forked it from

File: plugins/gatsby-source-jenkinsplugins/.gitignore
Comment (halkeye) [reply plugins/gatsby-source-jenkinsplugins/.gitignore]: Nope, left behind files from the source I forked it from

File: plugins/gatsby-source-jenkinsplugins/gatsby-node.js
Comment (halkeye) [reply plugins/gatsby-source-jenkinsplugins/gatsby-node.js]: yes... but i'd have to move some variables around, so lemme think about it.

**Comment by zbynek** [comment]
* When I get from search to a plugin and click the "<- Find plugins" button, it gets me to index. The current version of the site takes me to the search results instead. IMHO it makes no sense to duplicate functionality of the browser's back button, so it's fine to kill that feature, but the wording of the link should be different ("Browse Plugins", "All Plugins" or something like that)

* The icon view / detail view / list view switch on the search page doesn't work and is rendered strangely

File: src/components/Icon.jsx
Comment (halkeye) [reply src/components/Icon.jsx]: Yea, and not even the same logic, gross, fixing

File: src/components/Pages.jsx
Comment (halkeye) [reply src/components/Pages.jsx]: yea gross, i think i was just rushing at that point

File: src/styles/base.css
Comment (halkeye) [reply src/styles/base.css:870]: I think so, it was previously loading the entire css file every time, which was really messing everything up, so I just pulled in the ~6 lines it needed.

File: utils.js
Comment (halkeye) [reply utils.js]: Probably, but i like consistency?

File: README.md
Comment (halkeye) [reply README.md:10]: gatsby uses webpack, so i left it in

**Comment by halkeye** [comment]
> There's a deprecation warning when running `yarn dev`:

Yea thats wack, i'm not getting it. can you delete your dist directory and try again? I don't even have a dist directory



> How do I get the content to show up locally?
> 
> I tried with `SLOW_MODE` the only way I could get that to even finish was to add a hosts file entry, otherwise it kept crashing with random unknown hosts after a few minutes.
> 
> But even then there was no content

It does do a lot of http calls, one per plugin to get all the content. Your DNS server probably is getting slammed with traffic. You might want to setup a local caching proxy.

Testing the content thing now though.

**Comment by halkeye** [comment]
> * When I get from search to a plugin and click the "<- Find plugins" button, it gets me to index. IMHO it makes no sens to duplicate functionality of the browser's back button, so it's fine to kill that feature, but the wording of the link should be different ("Browse Plugins", "All Plugins" or something like that)

Agreed, I tried to keep the layout the same as previously, I don't really like the dialog look anyways.

> * The icon view / detail view / list view switch on the search page doesn't work and is rendered strangely

Oh yea, those... i forgot about them, I'll try to get that working this weekend.

**Comment by timja** [comment]
Mobiles looking better now, just the search box on the home page is squashed too far, you can only fit one or two letters before it’s off screen

**Comment by daniel-beck** [comment]
> It doesn't have to have the content inside the update center json, it could easily be seperate files

Like we do today for e.g. https://updates.jenkins-ci.org/current/plugin-documentation-urls.json

> Then different things could start consuming more and more of it, like inside jenkins could show the readme as an iframe or something.

I don't know whether there's a URL in GitHub that supports rendering the HTML content.

----

Generation of a "update center source for readmes" would likely need to be in a separate process, as we currently generate the update center every 4 minutes, and I worked hard to get it this low (from previously 20+ minutes). While GitHub may support caching and If-Updated-Since (or what it's called), the wiki probably doesn't in the way we're using it. GitHub API limits may also be a problem here.

**Comment by halkeye** [comment]
> Figure out deployment

For now I'm going to split the codebase into 2 repos. 
1) plugin-site-api docker image (blocked by https://issues.jenkins-ci.org/browse/INFRA-2400) 
2) Update helm chart to deploy both docker images independantly using the same ingress

**Comment by halkeye** [comment]
https://github.com/jenkins-infra/jenkins.io/pull/2731 should in theory allow me to try and pull all the required css into a single hosted file at build time, which should help first paint.

**Comment by halkeye** [comment]
@timja I'm thinking I'm happy with the mobile state right now

Anything other than the nice to haves need doing? I think its pretty done

**Comment by timja** [comment]
The home page is scrolling horizontally on mobile currently, although it does do this on existing and it looks better than before

Not that I’m aware of, just the helm chart I assume to go?

**Comment by halkeye** [comment]
I'll poke at that scrolling this afternoon

Waiting for a new docker repo to be created to work on the helm chart. I really think back and frontend should be seperate images

**Comment by halkeye** [comment]
> The home page is scrolling horizontally on mobile currently, although it does do this on existing and it looks better than before

Re-did the css for that and I think it looks better, but i'm horrible at spacing.

**Comment by timja** [comment]
You should be able to push it to your own docker repo for testing and override the image in a local values file.

**Comment by halkeye** [comment]
> You should be able to push it to your own docker repo for testing and override the image in a local values file.

Yea I guess I did that for wiki exporter, so i'll do so

**Comment by timja** [comment]
Mobile looks sooo much better now, great work

**Comment by timja** [comment]
If you click the browse button from the home page there’s a flicker of the filters and then they disappear and can’t reappear, there’s the down arrow to bring them back but it doesn’t work, likely fine would be good if it could lose the arrow on mobile.

The category links don’t work on the home page on mobile

The browse page appears to be missing a page title as well

**Comment by halkeye** [comment]
> If you click the browse button from the home page there’s a flicker of the filters and then they disappear and can’t reappear, there’s the down arrow to bring them back but it doesn’t work, likely fine would be good if it could lose the arrow on mobile.

can you screenshot it? There's no arrow for me at all, and the filters shouldn't be there on home page at all.

image


> The category links don’t work on the home page on mobile

yea, totally missed that

> The browse page appears to be missing a page title as well

good catch, and the main page still says "Jenkins Plugins | Jenkins Plugins" which I forgot to look into

**Comment by timja** [comment]
Other weird issue, if you use browse, click onto another page without any search and then try search something I.e blueocean no results will be found

**Comment by timja** [comment]
It’s on the browse page not the home page, screenshotted but not sure how to upload on mobile, lmk if you still need and I’ll upload somewhere

**Comment by halkeye** [comment]
> Other weird issue, if you use browse, click onto another page without any search and then try search something I.e blueocean no results will be found

I have absolutely no idea how to follow along. I click browse from the image page....
OH!
as in its not resetting back to page 1? ugh i hate that search box so much. Some things change live, some things wait for the submit button

**Comment by timja** [comment]
Browse => click page 2 => enter search term

Get no results, no pages are in the results
https://imgur.com/w1vFbrz

**Comment by halkeye** [comment]
> Browse => click page 2 => enter search term

I'm going to put that on todo later for now, cause its a bigger fix, and it only works for search text on the existing page

**Comment by timja** [comment]
Documentation seems missing on pages atm?
https://jenkins-plugins.g4v.dev/slack

Only the metadata is appearing

**Comment by halkeye** [comment]
> If you click the browse button from the home page there’s a flicker of the filters and then they disappear and can’t reappear, there’s the down arrow to bring them back but it doesn’t work, likely fine would be good if it could lose the arrow on mobile.

So its kinda by design, and kinda bad practice.

On existing site, the filters are taking up the first page and a half of the results screen. I felt that was kinda dumb, so moved the filters to after the results, (on the bottom of the screen), but I realize now that its also pretty dumb.

So I don't know what the right solution is.

They are there though, just after the results instead of before.

**Comment by timja** [comment]
This is the filter popup that you get on the browse page:
https://imgur.com/zdfOSHl

It appears on every page load on mobile while a search is happening

Could be a separate fix as well

**Comment by halkeye** [comment]
> This is the filter popup that you get on the browse page:
> https://imgur.com/zdfOSHl
> 
> It appears on every page load on mobile while a search is happening
> 
> Could be a separate fix as well

Yea, its because the spinner is positioned ontop of everything. If i get rid of the call to the plugins server, and just to it locally in browser, it would probably never happen.

ARG. I think i'm going to leave it.

> Documentation seems missing on pages atm?

Yep, my bad, apparently in my move to building inside of Docker, I forgot to keep pass the flag to actually grab the real content. It'll be back up soon

**Comment by halkeye** [comment]
> Documentation seems missing on pages atm?
> https://jenkins-plugins.g4v.dev/slack
> 
> Only the metadata is appearing

back

File: Jenkinsfile
Comment (olblak) [inline Jenkinsfile:37]: I guess this url will have to be updated

File: Jenkinsfile
Comment (olblak) [inline Jenkinsfile]: On trusted.ci.jenkins.io and ci.jenkins.io this line is not needed as they already have access to credentials.
ci.jenkins.io can push images to the jenkins4eval docker hub organisation (for testing purpose)
trusted.ci.jenkins.io can push to either jenkinsci/ jenkinsciinfra organisation

Comment (olblak) [reply Jenkinsfile]: An example [here](https://github.com/jenkins-infra/openvpn/blob/staging/Jenkinsfile#L51-L46) reusing a pipeline [library](https://github.com/jenkins-infra/pipeline-library/blob/master/vars/infra.groovy)

**Review by olblak — COMMENTED** [review]
:o 156 file changes 
On overall it's already a major improvement, removing ci.jenkins.io dependency and being able to search for plugin information on my phone is already awesome. I am still evaluating Netlify instead of azure.
I still have to give another review on this PR

File: Jenkinsfile
Comment (timja) [reply Jenkinsfile]: This pipeline will likely be modified before merge, it was discussed a bit in the docs sig gitter channel today, about not publishing a docker image and either using nginx with blob storage or a cdn like netlify

File: Dockerfile
Comment (timja) [inline Dockerfile]: is the intention to use nginx image? what's the reason for this comment being here?

**Comment by timja** [comment]
The homepage doesn't look great on my widescreen monitor:

Plugins index off center,
White bar between header and content
Logo position has changed not sure if thats an issue or not though

image

for comparison existing one:

image

**Comment by timja** [comment]
Is the text too far to the left on desktop now?

image

image

**Comment by timja** [comment]
The CSS seems to be missing for the pointer on the search page and the border highlight on hover is missing,

https://jenkins-plugins.g4v.dev/ui/search?query=azure

static site:
image

prod:
image

little hard to tell in the screenshot but their should be a "hand" pointer to indicate that the tile is clickable, and the border in the previous site was highlighted to further indicate which one would be clicked on I guess

**Comment by halkeye** [comment]
> Is the text too far to the left on desktop now?

thats not very actionable. Who should answer that? I can certainly give it more padding, but how much is good?

At some point maybe we should involve @josephbrueggen/ux sig

File: Dockerfile
Comment (halkeye) [reply Dockerfile]: left behind probably. the gatsby image is way faster than stock nginx, and i still think cdn > docker.

File: Jenkinsfile
Comment (halkeye) [reply Jenkinsfile]: @olblak these lines only run on my CI
```
      when {
        environment name: 'JENKINS_URL', value: 'https://jenkins.gavinmogan.com/'
      }
```
The other block will run things inside of infra push

File: Jenkinsfile
Comment (halkeye) [reply Jenkinsfile:37]: This is only the "test" server (in comments). Its running on my jenkins instance, everywhere else will just do the regular build.

**Comment by halkeye** [comment]
Okay, I think I've handled all the issues. Its getting super hard to keep track of change requests.

File: Jenkinsfile
Comment (timja) [inline Jenkinsfile]: Credential name needs updating or comment removed

Comment (timja) [reply Jenkinsfile]: Ah sorted already

Comment (halkeye) [reply Jenkinsfile]: .. i was very confused, I remember specifically removing that line and didn't think i committed it.

**Review by timja — APPROVED** [review]
LGTM, once the storage account is in place with creds on ci.jenkins.io we can ship I believe

File: Jenkinsfile
Comment (olblak) [inline Jenkinsfile]: Labels is important as we have linux  and windows machines with docker

Comment (olblak) [reply Jenkinsfile]: nevermind I got confused

Comment (halkeye) [reply Jenkinsfile]: I mean thats an outdated change. I switched it to docker agent. Is that still going to be an issue? how can I change it to docker agent and linux?

File: Jenkinsfile
Comment (olblak) [inline Jenkinsfile]: PLUGINSHTML_STORAGEACCOUNTKEYCan you rename  PLUGINSITE_STORAGEACCOUNTKEY in order to be consistent?

Comment (halkeye) [reply Jenkinsfile]: Done

Comment (olblak) [reply Jenkinsfile]: PLUGINSITE_STORAGEACCOUNTKEY is available on trusted.ci

Comment (halkeye) [reply Jenkinsfile]: Awesome. is there any way to test before merging? how about the other parameters for the uploader? i kinda guessed blindly

**Comment by timja** [comment]
> Awesome. is there any way to test before merging? how about the other parameters for the uploader? i kinda guessed blindly

🤷‍♂ 🚢

**Pull request merged**
